### PR TITLE
Add sync.Mutex to stop concurrent secret creation

### LINKED ISSUE #44347: [BUG] Stale secrets for impersonation user in downstream cluster
**Rancher Server Setup**
- Rancher version: 2.7.9, 2.8.2

**Information about the Cluster**
- Kubernetes version: v1.26.11+rke2r1
- Cluster Type (Local/Downstream): Downstream, created from Rancher on EC2

**User Information**
- What is the role of the user logged in?  `Admin`

**Describe the bug**

- Multiple stale `Secrets` created for a single impersonation user/ServiceAccount in the downstream cluster
- SA shows 1 secret and single reference but multiple secrets with same SA reference found
- The stale secrets don't get reconciled even if we delete the current referenced secret in the ServiceAccount
```
> k get secrets -n cattle-impersonation-system
NAME                                            TYPE                                  DATA   AGE
cattle-impersonation-u-hjxlz-token-24kwk        kubernetes.io/service-account-token   3      22h
cattle-impersonation-u-l767yhswxf-token-lnzc2   kubernetes.io/service-account-token   3      3d23h
cattle-impersonation-u-l767yhswxf-token-rdlkx   kubernetes.io/service-account-token   3      4d22h
cattle-impersonation-u-l767yhswxf-token-w5klt   kubernetes.io/service-account-token   3      3d23h
cattle-impersonation-u-oxfisb4qb2-token-gtjln   kubernetes.io/service-account-token   3      4d23h
cattle-impersonation-user-ptj5z-token-gzjnc     kubernetes.io/service-account-token   3      4d23h
> k get sa -n cattle-impersonation-system -o wide
NAME                                SECRETS   AGE
cattle-impersonation-u-hjxlz        1         22h
cattle-impersonation-u-l767yhswxf   1         4d22h
cattle-impersonation-u-oxfisb4qb2   1         4d23h
cattle-impersonation-user-ptj5z     1         4d23h
default                             0         4d23h
> k get secrets -A -o yaml | grep 'cattle-impersonation-u-l767yhswxf'
      kubernetes.io/service-account.name: cattle-impersonation-u-l767yhswxf
    generateName: cattle-impersonation-u-l767yhswxf-token-
    name: cattle-impersonation-u-l767yhswxf-token-lnzc2
      name: cattle-impersonation-u-l767yhswxf
      kubernetes.io/service-account.name: cattle-impersonation-u-l767yhswxf
    generateName: cattle-impersonation-u-l767yhswxf-token-
    name: cattle-impersonation-u-l767yhswxf-token-rdlkx
      name: cattle-impersonation-u-l767yhswxf
      kubernetes.io/service-account.name: cattle-impersonation-u-l767yhswxf
    generateName: cattle-impersonation-u-l767yhswxf-token-
    name: cattle-impersonation-u-l767yhswxf-token-w5klt
      name: cattle-impersonation-u-l767yhswxf
```

A snippet from the Rancher pod log shows that the multiple creation happens usually quickly after each other, hinting at lack of concurency while the Ensure func is called for the same ServiceAccount
```
2024/02/13 11:29:43 [INFO] EnsureSecretForServiceAccount: waiting for secret [cattle-impersonation-user-kgnqk-token-cd6fl] to be populated with token
2024/02/13 11:29:43 [INFO] EnsureSecretForServiceAccount: waiting for secret [cattle-impersonation-user-kgnqk-token-v7r6z] to be populated with token
2024/02/13 11:29:44 [INFO] EnsureSecretForServiceAccount: waiting for secret [cattle-impersonation-user-kgnqk-token-sg9sb] to be populated with token
2024/02/13 11:29:44 [INFO] Creating roleBinding User user-kgnqk Role admin in cattle-system
2024/02/13 11:29:44 [INFO] Updating clusterRole project-owner-promoted for project access to global resource.
2024/02/13 11:29:44 [INFO] [mgmt-auth-crtb-controller] Creating roleBinding for subject u-xpeyhn3op3 with role cluster-owner in namespace p-g8bfg
2024/02/13 11:29:44 [INFO] Creating roleBinding User user-kgnqk Role project-owner in default
2024/02/13 11:29:44 [INFO] [mgmt-auth-crtb-controller] Creating roleBinding for subject u-xpeyhn3op3 with role cluster-owner in namespace p-s2mp7
2024/02/13 11:29:44 [INFO] Creating roleBinding User user-kgnqk Role admin in kube-system
```
**To Reproduce**

This seems random at this point, on a Rancher with a downstream cluster, this can be seen out of the box in the downstream cluster.
A customer has seen this for a custom user, in my reproducer this is happening for a fleet user.
There's a time difference between the stale secrets creation, so this might not occur at the first try

### PR DESCRIPTION
**Pull request by vatsalparekh**
## Issue: 

Fixes #44347 
 
### Automated Testing

* Test types added/modified:
    * Unit
    * Integration (Go Framework)
    * Integration (v2prov Framework)
    * Validation (Go Framework)
    * Other - Explain: _EXPLAIN_
    * None
    * _REMOVE NOT APPLICABLE BULLET POINTS ABOVE_
* If "None" - Reason: _EXPLAIN THE REASON_
  
* If "None" - GH Issue/PR: _LINK TO GH ISSUE/PR TO ADD TESTS_

Summary: _TODO_

### Regressions Considerations

_TODO_

Existing / newly added automated tests that provide evidence there are no regressions:
* _TODO_

### CODE REVIEW

File: pkg/serviceaccounttoken/secret.go [hunk 1]
```diff
@@ -164,3 +188,57 @@ func isSecretForServiceAccount(secret *v1.Secret, sa *v1.ServiceAccount) bool {
...
+	leaseClient := clientSet.CoordinationV1().Leases(namespace)
+	err := leaseClient.Delete(ctx, fmt.Sprint(LeasePrefix+name), metav1.DeleteOptions{})
+	if err != nil {
+		return fmt.Errorf("error deleting lease: %w", err)
+	}
+	return nil
+}
+
+// Helper functions for getting pointers
+func stringPtr(s string) *string {
```
Comment (bigkevmcd) [inline pkg/serviceaccounttoken/secret.go:238]: You could replace these with a generic function:

```go
func valPtr[T any](v T) *T {
        return &v
}
```

Can we add tests for this change?

Comment (bigkevmcd) [reply pkg/serviceaccounttoken/secret.go:238]: UPDATE: I forgot about https://pkg.go.dev/k8s.io/utils/ptr

You can use `ptr.To()` from there.

Comment (vatsalparekh) [reply pkg/serviceaccounttoken/secret.go]: Fixed

**Review by bigkevmcd — DISMISSED** [review]
I've tested this heavily, and it looks good to me.

File: pkg/serviceaccounttoken/secret.go [hunk 2]
```diff
@@ -3,25 +3,35 @@ package serviceaccounttoken
...
 	// ServiceAccountSecretLabel is the label used to search for the secret belonging to a service account.
 	ServiceAccountSecretLabel = "cattle.io/service-account.name"
 
 	serviceAccountSecretAnnotation = "kubernetes.io/service-account.name"
+
+	// LeasePrefix is the name of the lease used to manage concurrency
+	LeasePrefix = "sa-token-lease-"
 )
 
+// Mutex to limit parellal writes by EnsureSecretForServiceAccount
```
Comment (enrichman) [inline pkg/serviceaccounttoken/secret.go:32]:
```suggestion
// Mutex to limit parallel writes by EnsureSecretForServiceAccount
```

**Review by enrichman — APPROVED** [review]
Just a small typo, otherwise it looks good to me.

**Comment by enrichman** [comment]
@samjustus I was waiting to approve again because of the failing build, but I've restarted Drone to see if it was a flaky one!
/cc @vatsalparekh

**Review by bigkevmcd — APPROVED** [review]

**Review by enrichman — APPROVED** [review]

**Pull request merged**